tools/frrcommon.sh.in

```diff
@@ -207,13 +207,39 @@ daemon_stop() {
 		return 1
 	fi
 
-	debug "kill -2 $pid"
-	kill -2 "$pid"
 	cnt=1200
+
+	#
+	# Allow calling program to convey the timeout it may be using
+	#
+	if [ -n "$FRR_WATCHFRR_TIMEOUT" ] ; then
+		$(( cnt = "$FRR_WATCHFRR_TIMEOUT" - 5 ))
+		$(( cnt *= 10 ))
+	fi
+
+	debug "kill -2 $pid, cnt $cnt"
+	log_success_msg "Stopping $dmninst, pid $pid ..."
+	kill -2 "$pid"
 	while kill -0 "$pid" 2>/dev/null; do
 		sleep .1
 		[ $(( cnt -= 1 )) -gt 0 ] || break
 	done
+
+	# Really kill the daemon if it didn't react to SIGINT
+	if kill -0 "$pid" 2>/dev/null; then
+		debug "kill -9 $dmninst, pid $pid"
+		log_success_msg "Failed to stop $dmninst, killing pid $pid ..."
+
+		kill -9 "$pid"
+
+		# And wait a bit for the kill to take effect
+		cnt=5
```

Comment:
Just curious how it should be under a high load, is this enough? :)

Reply:
I was concerned about (and saw) the case where the system was busy, and the "KILL" step took some time to have effect. If we just return "success" to watchfrr at this point, there's some risk that the old daemon will still be present. I was just trying to give a bit of an opportunity for the system to finish cleaning up the old process.

Reply:
@eqvinox @ton31337 did either of you have more feedback on this?  Is the approach okay as is, or do you have a suggestion on what you'd like to see changed before approving?